File: SpriteSheetUnpacker/ImageUnpacker.vb
#Region " Options "
Option Strict On
Option Explicit On
#End Region

Public Class ImageUnpacker
    Private original As Bitmap
    Private originalLock As Object = New Object()
    Private pcComplete As Integer = 0
    Private backgroundColour As Color? = Nothing
    Private boxesLock As Object = New Object()
    Private boxes As List(Of Rectangle)
    Private areaUnpacked As Integer
    Private areaUnpackedLock As Object = New Object()
    Private areaToUnpack As Integer
    Private threadCounter As Integer = 0
    Private threadCompleteCounterLock As Object = New Object()
    Private threadCompleteCounter As Integer = 0
    Private areAllThreadsCreated As Boolean
    Private isUnpackingComplete As Boolean = False
    Private originalSize As Size
    Private pallette As Imaging.ColorPalette = Nothing
    Private _isBackgroundColourSet As Boolean = False
    Private _isUnpacking As Boolean = False
    Private Const INT_MAX_REGION_WIDTH As Integer = 400    
    Public FileName As String
    Public ColoursCount As Integer = 0

    Public Event UnpackingComplete()
    Public Event PcCompleteChanged(ByVal pcComplete As Integer)

    Public Sub New(ByVal image As Bitmap, ByVal fileName As String)
        If image.Palette.Entries.Length > 0 Then
            Me.pallette = image.Palette
        End If
        Me.original = New Bitmap(CType(image.Clone, Bitmap))
        Me.originalSize = image.Size
        Me.boxes = New List(Of Rectangle)
        Me.FileName = fileName
    End Sub

    Public Function GetPallette() As Imaging.ColorPalette
        Return Me.pallette
    End Function

    Public Function GetSize() As Size
        Return Me.originalSize
    End Function

    Public Function IsUnpacking() As Boolean
        Return Me._isUnpacking
    End Function

    Public Function IsBackgroundColourSet() As Boolean
        Return Me._isBackgroundColourSet
    End Function

    Public Function GetBackgroundColour() As Color
        If Me.backgroundColour.HasValue Then
            Return Me.backgroundColour.Value
        Else
            Return Color.Black
        End If
    End Function

    Public Function GetBoxes() As List(Of Rectangle)
        Return New List(Of Rectangle)(Me.boxes)
    End Function

    Public Function IsUnpacked() As Boolean
        Return Me.isUnpackingComplete
    End Function

    Public Function GetPcComplete() As Integer
        Return Me.pcComplete
    End Function

    Public Function GetOriginalClone() As Bitmap
        Dim clone As Bitmap

        SyncLock (Me.originalLock)
            clone = New Bitmap(CType(Me.original.Clone(), Bitmap))
        End SyncLock

        Return clone
    End Function

    Private Sub SetPcComplete(ByVal pcComplete As Integer)
        If pcComplete > Me.pcComplete Then
            Me.pcComplete = pcComplete
            RaiseEvent PcCompleteChanged(Me.pcComplete)
        End If
    End Sub

    Public Sub StartUnpacking()
        Dim newThread As Threading.Thread = New Threading.Thread(AddressOf Me.Unpack)
        Me.isUnpackingComplete = False
        Me.pcComplete = 0
        Me.boxes.Clear()
        Me._isUnpacking = True
        Me.threadCounter = 0
        Me.threadCompleteCounter = 0
        newThread.Start()
    End Sub

    Public Shared Function OrderBoxes(ByVal boxes As List(Of Rectangle), ByVal enuSelectAllOrder As SelectAllOrder, ByVal spriteSheetSize As Size) As List(Of Rectangle)
        Dim colOrderedFrames As SortedDictionary(Of Integer, List(Of Rectangle)) = New SortedDictionary(Of Integer, List(Of Rectangle))
        Dim intLocation As Integer
        Dim returnedOrder As List(Of Rectangle) = New List(Of Rectangle)

        For Each objFrame As Rectangle In boxes

            Select Case enuSelectAllOrder
                Case SelectAllOrder.TopLeft
                    intLocation = objFrame.X + (objFrame.Y * spriteSheetSize.Width)
                Case SelectAllOrder.BottomLeft
                    intLocation = objFrame.X + ((objFrame.Y + objFrame.Height) * spriteSheetSize.Width)
                Case SelectAllOrder.Centre
                    intLocation = CInt((objFrame.X + (objFrame.Width / 2)) + ((objFrame.Y + (objFrame.Height / 2)) * spriteSheetSize.Width))
            End Select

            If Not colOrderedFrames.ContainsKey(intLocation) Then
                colOrderedFrames.Add(intLocation, New List(Of Rectangle))
            End If
            colOrderedFrames(intLocation).Add(objFrame)
        Next

        For Each intLocationKey As Integer In colOrderedFrames.Keys
            For Each objFrame As Rectangle In colOrderedFrames(intLocationKey)
                returnedOrder.Add(objFrame)
            Next
        Next

        Return returnedOrder
    End Function

    Private Sub Unpack(ByVal state As Object)
        Try
            Dim intXSize As Integer
            Dim intYSize As Integer
            Dim region As Rectangle
            Dim regionThread As Threading.Thread

            Me.areAllThreadsCreated = False
            If Not Me.backgroundColour.HasValue Then
                Me.SetBackgroundColour(Me.GetOriginalClone())
            End If
            Me.SetPcComplete(10)

            intYSize = CInt(Math.Ceiling(Me.originalSize.Height / 3))
            intXSize = CInt(Math.Ceiling(Me.originalSize.Width / 3))

            Me.areaToUnpack = Me.originalSize.Width * Me.originalSize.Height

            For y As Integer = 0 To intYSize * 4 Step intYSize
                For x As Integer = 0 To intXSize * 4 Step intXSize

                    region = New Rectangle(x, y, Math.Min(intXSize, (Me.originalSize.Width - x) - 1), Math.Min(intYSize, (Me.originalSize.Height - y) - 1))
                    regionThread = New Threading.Thread(AddressOf Me.HandleDividedAreaThread)
                    regionThread.Name = "Region thread " & (y * (intXSize * 4)) + x
                    Me.threadCounter += 1
                    regionThread.Start(region)

                    Me.SetPcComplete(10 + CInt((((Math.Min(y, Me.originalSize.Height) * (Me.originalSize.Width - 1)) + Math.Min(x, Me.originalSize.Width)) / (Me.originalSize.Height * Me.originalSize.Width)) * 10))
                Next
            Next

            Me.SetPcComplete(20)
            Me.areAllThreadsCreated = True
            SyncLock (Me.threadCompleteCounterLock)
                If Me.threadCompleteCounter = Me.threadCounter Then
                    Me.HandleUnpackComplete()
                End If
            End SyncLock
        Catch ex As Exception
            ForkandBeard.Logic.ExceptionHandler.HandleException(ex, "cat@forkandbeard.co.uk")
        End Try
    End Sub

    Private Sub HandleDividedAreaThread(ByVal regionObject As Object)
        Try
            Me.HandleDividedArea(CType(regionObject, Rectangle), True, Me.GetOriginalClone())
        Catch ex As Exception
            ForkandBeard.Logic.ExceptionHandler.HandleException(ex, "cat@forkandbeard.co.uk")
        End Try
    End Sub

    Private Sub HandleDividedArea(ByVal region As Rectangle, ByVal updateCounter As Boolean, ByVal image As Bitmap)

        If region.Width > INT_MAX_REGION_WIDTH _
        OrElse region.Height > INT_MAX_REGION_WIDTH Then
            Dim quarterRegions As List(Of Rectangle) = New List(Of Rectangle)

            ' Top left.
            quarterRegions.Add(New Rectangle(region.X, region.Y, CInt(region.Width / 2), CInt(region.Height / 2)))
            ' Top right.
            quarterRegions.Add(New Rectangle(region.X + CInt(region.Width / 2), region.Y, CInt(region.Width / 2) + 1, CInt(region.Height / 2)))
            ' Bottom left.
            quarterRegions.Add(New Rectangle(region.X, region.Y + CInt(region.Height / 2), CInt(region.Width / 2), CInt(region.Height / 2) + 1))
            ' Bottom right.
            quarterRegions.Add(New Rectangle(region.X + CInt(region.Width / 2), region.Y + CInt(region.Height / 2), CInt(region.Width / 2) + 1, CInt(region.Height / 2) + 1))
            For Each quarter As Rectangle In quarterRegions
                Me.HandleDividedArea(quarter, False, image)
            Next
        Else
            Using unpacker As RegionUnpacker = New RegionUnpacker(image, region, Me.backgroundColour.Value)
                unpacker.UnpackRegion()
                SyncLock (Me.areaUnpackedLock)
                    Me.areaUnpacked += CInt((region.Width * region.Height) * 0.8)
                End SyncLock

                Me.SetPcComplete(20 + CInt((Me.areaUnpacked / Me.areaToUnpack) * 75))

                SyncLock (Me.boxesLock)
                    Me.boxes.AddRange(unpacker.Boxes)
                    'RegionUnpacker.CombineBoxes(Me.boxes, Me.backgroundColour.Value, unpacker.GetImage())
                End SyncLock
            End Using

            SyncLock (Me.areaUnpackedLock)
                Me.areaUnpacked += CInt((region.Width * region.Height) * 0.2)
            End SyncLock

            Me.SetPcComplete(20 + CInt((Me.areaUnpacked / Me.areaToUnpack) * 80))
        End If

        If updateCounter Then
            SyncLock (Me.threadCompleteCounterLock)
                Me.threadCompleteCounter += 1
                If Me.areAllThreadsCreated _
                AndAlso Me.threadCompleteCounter = Me.threadCounter Then
                    Me.HandleUnpackComplete()
                End If
            End SyncLock
        End If
    End Sub

    Private Sub SetBackgroundColour(ByVal image As Bitmap)
        Dim colours As Dictionary(Of Color, Integer) = New Dictionary(Of Color, Integer)
        Dim presentColour As Color
        Dim maxCount As Integer = 0

        For x As Integer = 0 To Me.originalSize.Width - 1
            For y As Integer = 0 To Me.originalSize.Height - 1
                presentColour = image.GetPixel(x, y)

                If Not colours.ContainsKey(presentColour) Then
                    colours.Add(presentColour, 1)
                Else
                    colours(presentColour) += 1
                End If

                If (x + y) Mod 100 = 0 Then
                    Dim intTotal As Integer

                    intTotal = Me.originalSize.Width * Me.originalSize.Height
                    Me.SetPcComplete(CInt((((x * (Me.originalSize.Height - 1)) + y) / intTotal) * 10))
                End If
            Next
        Next

        For Each colour As Color In colours.Keys
            If colours(colour) >= maxCount Then
                maxCount = colours(colour)
                Me.backgroundColour = colour
            End If
        Next
        image.Dispose()
        Me._isBackgroundColourSet = True
        Me.ColoursCount = colours.Count - 1
    End Sub

    Private Sub HandleUnpackComplete()
        Using img As Bitmap = Me.GetOriginalClone()
            SyncLock (Me.boxesLock)
                RegionUnpacker.CombineBoxes(Me.boxes, Me.backgroundColour.Value, img)
            End SyncLock
        End Using

        Me.isUnpackingComplete = True
        Me._isUnpacking = False
        Me.SetPcComplete(100)
        RaiseEvent UnpackingComplete()
    End Sub
End Class

File: ASU/BO/ImageUnpacker.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Drawing;

namespace ASU.BO
{
    public class ImageUnpacker
    {
        public bool IsLarge { get; set; }

        private Bitmap original;
        private object originalLock = new object();
        private int pcComplete = 0;
        private Color? backgroundColour = null;
        private object boxesLock = new object();
        private List<Rectangle> boxes;
        private int areaUnpacked;
        private object areaUnpackedLock = new object();
        private int areaToUnpack;
        private int threadCounter = 0;
        private object threadCompleteCounterLock = new object();
        private int threadCompleteCounter = 0;
        private bool areAllThreadsCreated;
        private bool isUnpackingComplete = false;
        private Size originalSize;
        private System.Drawing.Imaging.ColorPalette pallette = null;
        private bool _isBackgroundColourSet = false;
        private bool _isUnpacking = false;
        private const int INT_MAX_REGION_WIDTH = 400;

        public string FileName { get; set; }
        public int ColoursCount = 0;

        public event UnpackingCompleteEventHandler UnpackingComplete;        
        public event PcCompleteChangedEventHandler PcCompleteChanged;

        public delegate void PcCompleteChangedEventHandler(int pcComplete);
        public delegate void UnpackingCompleteEventHandler();

        public ImageUnpacker(Bitmap image, string fileName, bool removeTransparency)
        {
            if (image.Palette.Entries.Length > 0)
            {
                this.pallette = image.Palette;
            }
            this.original = new Bitmap((Bitmap)image.Clone());
            this.originalSize = image.Size;
            if (removeTransparency)
            {
                this.original = this.RemoveTransparencyFromImage(this.original);
            }
            this.boxes = new List<Rectangle>();
            this.FileName = fileName;
            this.IsLarge = (this.original.Width * this.original.Height) > (800 * 800);
        }

        private Bitmap RemoveTransparencyFromImage(Bitmap image)
        {
            Dictionary<int, Color> coloursByArgb = new Dictionary<int, Color>();
            Dictionary<int, Color> transparentPixelsByCoord = new Dictionary<int, Color>();
            Dictionary<Color, Color> opaquedByTransparent = new Dictionary<Color, Color>();
            Color pixel;
            bool containsTransparency = false;
            Color transparentPixel;
            Color opaquedPixel;

            for(int x = 0; x < image.Width; x++)
            {
                for(int y = 0; y < image.Height; y++)
                {
                    pixel = image.GetPixel(x, y);

                    if (!coloursByArgb.ContainsKey(pixel.ToArgb()))
                    {
                        coloursByArgb.Add(pixel.ToArgb(), pixel);     
                    }
                    if (pixel.A < 255)
                    {
                        containsTransparency = true;
                        transparentPixelsByCoord.Add(GetHashFromCoord(new Point(x, y)), pixel);
                    }
                }
            }

            if (containsTransparency)
            {
                foreach (int coordHash in transparentPixelsByCoord.Keys)
                {
                    Point coord = GetCoordFromHash(coordHash);

                    transparentPixel = transparentPixelsByCoord[coordHash];
                    if (opaquedByTransparent.ContainsKey(transparentPixel))
                    {
                        opaquedPixel = opaquedByTransparent[transparentPixel];
                    }
                    else
                    {
                        opaquedPixel = Color.FromArgb(255, transparentPixel);

                        do
                        {
                            if (transparentPixel.R > (255 / 2))
                            {
                                opaquedPixel = Color.FromArgb(opaquedPixel.R - 1, opaquedPixel.G, opaquedPixel.B);
                            }
                            else
                            {
                                opaquedPixel = Color.FromArgb(opaquedPixel.R + 1, opaquedPixel.G, opaquedPixel.B);
                            }
                        } while (coloursByArgb.ContainsKey(opaquedPixel.ToArgb()));

                        coloursByArgb.Remove(transparentPixel.ToArgb());
                        coloursByArgb.Add(opaquedPixel.ToArgb(), opaquedPixel);
                        opaquedByTransparent.Add(transparentPixel, opaquedPixel);
                    }
                    image.SetPixel(coord.X, coord.Y, opaquedPixel);
                }
            }

            return image;
        }

        public System.Drawing.Imaging.ColorPalette GetPallette()
        {
            return this.pallette;
        }

        public Size GetSize()
        {
            return this.originalSize;
        }

        public bool IsUnpacking()
        {
            return this._isUnpacking;
        }

        public bool IsBackgroundColourSet()
        {
            return this._isBackgroundColourSet;
        }

        public Color GetBackgroundColour()
        {
            if (this.backgroundColour.HasValue)
            {
                return this.backgroundColour.Value;
            }
            else
            {
                return Color.Black;
            }
        }

        public List<Rectangle> GetBoxes()
        {
            return new List<Rectangle>(this.boxes);
        }

        public bool IsUnpacked()
        {
            return this.isUnpackingComplete;
        }

        public int GetPcComplete()
        {
            return this.pcComplete;
        }

        public Bitmap GetOriginalClone()
        {
            Bitmap clone = default(Bitmap);

            lock ((this.originalLock))
            {
                clone = new Bitmap((Bitmap)this.original.Clone());
            }

            return clone;
        }

        private void SetPcComplete(int pcComplete)
        {
            if(pcComplete > 100)
            {   // TODO: Fix this bug.
                pcComplete = 100;
            }
            if (pcComplete > this.pcComplete)
            {
                this.pcComplete = pcComplete;
                if (PcCompleteChanged != null)
                {
                    PcCompleteChanged(this.pcComplete);
                }
            }
        }

        public void StartUnpacking()
        {
            System.Threading.Thread newThread = new System.Threading.Thread(this.Unpack);
            this.isUnpackingComplete = false;
            this.pcComplete = 0;
            this.boxes.Clear();
            this._isUnpacking = true;
            this.threadCounter = 0;
            this.threadCompleteCounter = 0;
            newThread.IsBackground = true;
            newThread.Start();
        }

        public static List<Rectangle> OrderBoxes(List<Rectangle> boxes, Enums.SelectAllOrder selectAllOrder, Size spriteSheetSize)
        {
            SortedDictionary<int, List<Rectangle>> orderedBoxes = new SortedDictionary<int, List<Rectangle>>();
            int location = 0;
            List<Rectangle> returnedOrder = new List<Rectangle>();


            foreach (Rectangle box in boxes)
            {
                switch (selectAllOrder)
                {
                    case Enums.SelectAllOrder.TopLeft:
                        location = box.X + (box.Y * spriteSheetSize.Width);
                        break;
                    case Enums.SelectAllOrder.BottomLeft:
                        location = box.X + ((box.Y + box.Height) * spriteSheetSize.Width);
                        break;
                    case Enums.SelectAllOrder.Centre:
                        location = Convert.ToInt32((box.X + (box.Width / 2)) + ((box.Y + (box.Height / 2)) * spriteSheetSize.Width));
                        break;
                }

                if (!orderedBoxes.ContainsKey(location))
                {
                    orderedBoxes.Add(location, new List<Rectangle>());
                }
                orderedBoxes[location].Add(box);
            }

            foreach (int locationKey in orderedBoxes.Keys)
            {
                foreach (Rectangle box in orderedBoxes[locationKey])
                {
                    returnedOrder.Add(box);
                }
            }

            return returnedOrder;
        }

        private void Unpack(object state)
        {
            try
            {
                int subRegionCount;
                int xSize = 0;
                int ySize = 0;
                decimal totalSize;
                Rectangle region = default(Rectangle);
                List<System.Threading.Thread> regionThreads = new List<System.Threading.Thread>();
                List<Rectangle> regions = new List<Rectangle>();
                System.Threading.Thread regionThread = null;

                this.areAllThreadsCreated = false;
                if (!this.backgroundColour.HasValue)
                {
                    this.SetBackgroundColour(this.GetOriginalClone());
                }
                this.SetPcComplete(10);

                if(Environment.ProcessorCount > 1)
                {
                    subRegionCount = 4;
                }
                else
                {
                    subRegionCount = 2;
                }

                ySize = Convert.ToInt32(Math.Ceiling((double)this.originalSize.Height / (double)subRegionCount));
                xSize = Convert.ToInt32(Math.Ceiling((double)this.originalSize.Width / (double)subRegionCount));

                this.areaToUnpack = this.originalSize.Width * this.originalSize.Height;

                totalSize = this.originalSize.Height * this.originalSize.Width;
                
                for (int y = 0; y < this.originalSize.Height; y += ySize)
                {
                    for (int x = 0; x < this.originalSize.Width; x += xSize)
                    {
                        region = new Rectangle(x, y, Math.Min(xSize+1, (this.originalSize.Width - x) - 1), Math.Min(ySize+1, (this.originalSize.Height - y) - 1));
                        regionThread = new System.Threading.Thread(this.HandleDividedAreaThread);
                        regionThread.Name = "Region thread " + (y * (xSize * 4)) + x;
                        this.threadCounter += 1;
                        regionThread.IsBackground = true;
                        
                        regionThreads.Add(regionThread);
                        regions.Add(region);                        
                    }
                }

                for (int k = 0; k < regionThreads.Count; k++)
                {
                    regionThreads[k].Start(regions[k]);
                    this.SetPcComplete(10 + Convert.ToInt32((((Math.Min(regions[k].Y, this.originalSize.Height) * (this.originalSize.Width - 1)) + Math.Min(regions[k].X, this.originalSize.Width)) / totalSize) * 10));
                }

                this.SetPcComplete(20);
                this.areAllThreadsCreated = true;
                lock ((this.threadCompleteCounterLock))
                {
                    if (this.threadCompleteCounter == this.threadCounter)
                    {
                        this.HandleUnpackComplete();
                    }
                }
            }
            catch (Exception ex)
            {
                ForkandBeard.Logic.ExceptionHandler.HandleException(ex, "cat@forkandbeard.co.uk");
            }
        }

        private void HandleDividedAreaThread(object regionObject)
        {
            try
            {
                if (Environment.ProcessorCount == 1)
                {
                    System.Threading.Thread.Sleep(25);
                }

                this.HandleDividedArea((Rectangle)regionObject, true, this.GetOriginalClone());
            }
            catch (Exception ex)
            {
                ForkandBeard.Logic.ExceptionHandler.HandleException(ex, "cat@forkandbeard.co.uk");
            }
        }


        private void HandleDividedArea(Rectangle region, bool updateCounter, Bitmap image)
        {
            if (region.Width > INT_MAX_REGION_WIDTH || region.Height > INT_MAX_REGION_WIDTH)
            {
                List<Rectangle> quarterRegions = new List<Rectangle>();

                // Top left.
                quarterRegions.Add(new Rectangle(region.X, region.Y, Convert.ToInt32(region.Width / 2), Convert.ToInt32(region.Height / 2)));
                // Top right.
                quarterRegions.Add(new Rectangle(region.X + Convert.ToInt32(region.Width / 2), region.Y, Convert.ToInt32(region.Width / 2) + 1, Convert.ToInt32(region.Height / 2)));
                // Bottom left.
                quarterRegions.Add(new Rectangle(region.X, region.Y + Convert.ToInt32(region.Height / 2), Convert.ToInt32(region.Width / 2), Convert.ToInt32(region.Height / 2) + 1));
                // Bottom right.
                quarterRegions.Add(new Rectangle(region.X + Convert.ToInt32(region.Width / 2), region.Y + Convert.ToInt32(region.Height / 2), Convert.ToInt32(region.Width / 2) + 1, Convert.ToInt32(region.Height / 2) + 1));
                foreach (Rectangle quarter in quarterRegions)
                {
                    this.HandleDividedArea(quarter, false, image);
                }
            }
            else
            {
                using (RegionUnpacker unpacker = new RegionUnpacker(image, region, this.backgroundColour.Value))
                {
                    unpacker.UnpackRegion();
                    lock ((this.areaUnpackedLock))
                    {
                        this.areaUnpacked += Convert.ToInt32((double)(region.Width * region.Height) * 0.8f);
                    }

                    this.SetPcComplete(20 + Convert.ToInt32(((double)this.areaUnpacked / (double)this.areaToUnpack) * 75f));

                    lock ((this.boxesLock))
                    {
                        this.boxes.AddRange(unpacker.Boxes);
                    }
                }

                lock ((this.areaUnpackedLock))
                {
                    this.areaUnpacked += Convert.ToInt32((double)(region.Width * region.Height) * 0.2f);
                }

                this.SetPcComplete(20 + Convert.ToInt32(((double)this.areaUnpacked / (double)this.areaToUnpack) * 80f));
            }

            if (updateCounter)
            {
                lock ((this.threadCompleteCounterLock))
                {
                    this.threadCompleteCounter += 1;
                    if (this.areAllThreadsCreated && this.threadCompleteCounter == this.threadCounter)
                    {
                        this.HandleUnpackComplete();
                    }
                }
            }
        }

        private void SetBackgroundColour(Bitmap image)
        {
            Dictionary<int, int> colourCountsByArgb = new Dictionary<int, int>();
            Color presentColour; 
            int maxCount = 0;

            for (int x = 0; x <= this.originalSize.Width - 1; x++)
            {
                for (int y = 0; y <= this.originalSize.Height - 1; y++)
                {
                    presentColour = image.GetPixel(x, y);

                    if (!colourCountsByArgb.ContainsKey(presentColour.ToArgb()))
                    {
                        colourCountsByArgb.Add(presentColour.ToArgb(), 1);
                    }
                    else
                    {
                        colourCountsByArgb[presentColour.ToArgb()] += 1;
                    }

                    if ((x + y) % 100 == 0)
                    {
                        decimal total = 0;

                        total = this.originalSize.Width * this.originalSize.Height;
                        this.SetPcComplete(Convert.ToInt32((((x * (this.originalSize.Height - 1)) + y) / total) * 10));
                    }
                }
            }

            foreach (int colourArgb in colourCountsByArgb.Keys)
            {
                if (colourCountsByArgb[colourArgb] >= maxCount)
                {
                    maxCount = colourCountsByArgb[colourArgb];
                    this.backgroundColour = Color.FromArgb(colourArgb);
                }
            }
            image.Dispose();
            this._isBackgroundColourSet = true;
            this.ColoursCount = colourCountsByArgb.Count - 1;
        }

        private void HandleUnpackComplete()
        {
            using (Bitmap img = this.GetOriginalClone())
            {
                lock ((this.boxesLock))
                {
                    RegionUnpacker.CombineBoxes(ref this.boxes, this.backgroundColour.Value, img);
                }
            }

            this.isUnpackingComplete = true;
            this._isUnpacking = false;
            this.SetPcComplete(100);
            if (UnpackingComplete != null)
            {
                UnpackingComplete();
            }
        }

        private Point GetCoordFromHash(int hash)
        {
            return new Point((int)(hash >> 16), (int)(hash & 0x0000FFFF));
        }

        private int GetHashFromCoord(Point point)
        {
            return (point.X << 16) + point.Y;
        }
    }
}
